### PR TITLE
support sweeping multiline option in AutoML

### PR DESCRIPTION
**Pull request by LittleLittleCloud**

### CODE REVIEW

File: src/Microsoft.ML.AutoML/ColumnInference/TextFileContents.cs [hunk 1]
```diff
@@ -50,12 +50,14 @@ public static ColumnSplitResult TrySplitColumns(MLContext context, IMultiStreamS
...
             var sparse = new[] { false, true };
             var quote = new[] { true, false };
+            var tryMultiline = new[] { false, true };
             var foundAny = false;
             var result = default(ColumnSplitResult);
             foreach (var perm in (from _allowSparse in sparse
                                     from _allowQuote in quote
                                     from _sep in separatorCandidates
-                                    select new { _allowSparse, _allowQuote, _sep }))
+                                    from _tryMultiline in tryMultiline
```
Comment (justinormont) [inline src/Microsoft.ML.AutoML/ColumnInference/TextFileContents.cs:59]: naming pattern would be:
```suggestion
                                    from _allowMultiline in multiline
```

File: test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+""""
+
+""
+
+""
+
+",1,2
+
+
+
```
Comment (justinormont) [inline test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt:36]: There a a variety of other places that the multi-line needs to be added.

Basically anywhere "sparse" is within the AutoML/CodeGen:
* https://github.com/dotnet/machinelearning/search?q=sparse+path%3Asrc%2FMicrosoft.ML.AutoML&unscoped_q=sparse+path%3Asrc%2FMicrosoft.ML.AutoML
* https://github.com/dotnet/machinelearning/search?q=sparse+path%3Asrc%2FMicrosoft.ML.CodeGenerator&unscoped_q=sparse+path%3Asrc%2FMicrosoft.ML.CodeGenerator

File: test/Microsoft.ML.AutoML.Tests/ColumnInferenceTests.cs [hunk 3]
```diff
@@ -186,5 +187,18 @@ public void InferColumnsColumnInfoParam()
             Assert.Equal(DefaultColumnNames.Features, result.ColumnInformation.NumericColumnNames.First());
             Assert.Null(result.ColumnInformation.ExampleWeightColumnName);
         }
+
+        [Fact]
+        public void TrySplitColumns_should_split_on_dataset_with_newline_between_double_quotes()
+        {
```
Comment (antoniovs1029) [inline test/Microsoft.ML.AutoML.Tests/ColumnInferenceTests.cs:193]: _(pinned comment to code to allow threaded discussion)_

The original issue for this (#4460) mentioned problems with using files with new line characters inside quoted fields, when using ColumnInference() in this way:
```
MLContext mlContext = new MLContext();
var inputColumnInformation = new ColumnInformation();
inputColumnInformation.LabelColumnName = @"review_scores_rating";
var train = mlContext.Auto().InferColumns(TrainDataPath, inputColumnInformation);
```
So please, add a test actually testing this scenario of calling InferColumns() directly. I was under the impression that to fix this scenario other changes needed to be done in the ColumnInference internal API:
https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.AutoML/ColumnInference/ColumnInferenceApi.cs

File: test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt [hunk 4]
```diff
@@ -0,0 +1,36 @@
+﻿id,Column1,Column2,Column3
+1,this is a description, 1,2
+2,"this is a quote description",1,2
```
Comment (antoniovs1029) [inline test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt:3]: _(pinned to unrelated code to allow threaded discussion)_

@harishsk and I were under the impression that in order for ModelBuilder to support the new readMultilines option, then some changes might be needed on the CodeGenerator, as now the code generated by ModelBuilder should also include `readMultilines = true` if it was found necessary to use that parameter.

Are these changes not necessary?

Comment (LittleLittleCloud) [reply test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt:3]: Yes and no... CodeGen project should only refer to ML.Net packages that are already published. That's why we want to pin CodeGen to a specific version of ML.Net to avoid  any API changes breaking CodeGen. So from this aspect we need to upgrade CodeGen. Meanwhile we can't change CodeGen before the release. Because if we update CodeGen, we need to refer to a nightly build ML.Net for generating project, which is not available externally.

My suggestion is to not update CodeGen this time, becausse it won't affect a lot. CodeGenerated project uses LoadFromTextFile API with user-defined ModelInput Class, which should work even if there's newline between quotes if I recall correctly.

File: test/Microsoft.ML.AutoML.Tests/ColumnInferenceTests.cs (hunk 3 above)
Comment (LittleLittleCloud) [reply test/Microsoft.ML.AutoML.Tests/ColumnInferenceTests.cs:193]: ### Resolved

File: test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt (hunk 2 above)
Comment (LittleLittleCloud) [reply test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt:36]: ### Resolved

File: test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt (hunk 4 above)
Comment (antoniovs1029) [reply test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt:3]: I see, then it's ok.

> CodeGenerated project uses LoadFromTextFile API with user-defined ModelInput Class, which should work even if there's newline between quotes if I recall correctly.

Using the generic LoadFromTextFile<ModelInput> API you mention will still use the default value for readMultilines, which is set to false. It would be needed to actually set it explicitly to true if it was necessary.

Notice that up until recently LoadFromTextFile<ModelInput> only accepted some parameters on its signature, and I couldn't add the new readMultilines parameter there because that would be a breaking API change (which we are not allowed to do until ML.NET version 2). So instead, what I did was to add a new LoadFromTextFile<ModelInput>(path, options) overload, where it accepts a TextLoader.Options object (see #5134).

So if users wanted to use the generic LoadFromTextFile<> and also use the new options (such as readMultilines, or the ones being introduced in #5147 or #5145) then they will have to use the new public overload. In this sense, I think that CodeGenerator should also start using this other overload to have access to new options.

File: test/Microsoft.ML.AutoML.Tests/ColumnInferenceTests.cs [hunk 5]
```diff
@@ -200,5 +200,26 @@ public void TrySplitColumns_should_split_on_dataset_with_newline_between_double_
...
+            // Check if we can infer the column information
+            // from and input file which has escaped newlines inside quotes
+            var dataPath = GetDataPath("multiline.csv");
+            MLContext mlContext = new MLContext();
+            var inputColumnInformation = new ColumnInformation();
+            inputColumnInformation.LabelColumnName = @"id";
+            var result = mlContext.Auto().InferColumns(dataPath, inputColumnInformation);
+
+            // File only have 3 columns: "id", "description" and "animal"
```
Comment (justinormont) [inline test/Microsoft.ML.AutoML.Tests/ColumnInferenceTests.cs:215]:
```suggestion
            // File has 3 columns: "id", "description" and "animal"
```

File: src/Microsoft.ML.AutoML/ColumnInference/TextFileContents.cs [hunk 6]
```diff
@@ -23,14 +23,16 @@ public class ColumnSplitResult
 
             public bool AllowQuote { get; set; }
             public bool AllowSparse { get; set; }
+            public bool  ReadMultilines { get; set; }
 
-            public ColumnSplitResult(bool isSuccess, char? separator, bool allowQuote, bool allowSparse, int columnCount)
+            public ColumnSplitResult(bool isSuccess, char? separator, bool allowQuote, bool readMultilines, bool allowSparse, int columnCount)
```
Comment (justinormont) [inline src/Microsoft.ML.AutoML/ColumnInference/TextFileContents.cs:28]: Following naming convention would be:
```suggestion
            public ColumnSplitResult(bool isSuccess, char? separator, bool allowQuote, bool allowMultiline, bool allowSparse, int columnCount)
```

File: test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt (hunk 2 above)
Comment (justinormont) [reply test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt:36]: There's still some missing, like:
https://github.com/dotnet/machinelearning/blob/c0252704b3b54ebff20c5333fb449ba38ffaed92/src/Microsoft.ML.AutoML/ColumnInference/ColumnInferenceApi.cs#L31-L32

Comment (LittleLittleCloud) [reply test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt:36]: I prefer not exposing readMultiline option in  these APIs, Simply sweeping it in Infersplit  should be enough

File: test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt (hunk 4 above)
Comment (justinormont) [reply test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt:3]: Unless I'm missing part of it, if the CLI / Model Builder calls this new version of AutoML, you will have to update CodeGen. 

Otherwise the multi-line files will be read by the AutoML code, but the CodeGen'd code will not read the user's file.

> CodeGen project should only refer to ML.Net packages that are already published.

All of `Microsoft.ML.*` should be self consistent with each other. No package pinning needed. Almost all changes to the AutoML code should have a mirroring change to the CodeGen code. 

When I import `Microsoft.ML.CodeGeneration@0.17.0` into my project, I'm expecting it to reference `Microsoft.ML@1.5.0` & `Microsoft.ML.AutoML@0.17.0`. Lets say my project has `Microsoft.ML@1.5.0` to do prediction, then I want to call CodeGen (yes, only an internal call currently). Then my project needs both `Microsoft.ML@1.5.0` and `Microsoft.ML@1.4.x`, causing a [version conflict](https://docs.microsoft.com/en-us/nuget/concepts/dependency-resolution#cousin-dependencies). Consistent releases are better.

> Meanwhile we can't change CodeGen before the release. Because if we update CodeGen, we need to refer to a nightly build ML.Net for generating project, which is not available externally.

What's stopping CodeGen from being updated before the release? Why would a change to CodeGen cause CLI/ModelBuilder to depend on a nightly build instead of the released version?

Comment (LittleLittleCloud) [reply test/Microsoft.ML.AutoML.Tests/TestData/DatasetWithNewlineBetweenQuotes.txt:3]: > What's stopping CodeGen from being updated before the release? Why would a change to CodeGen cause CLI/ModelBuilder to depend on a nightly build instead of the released version?

It's not CLI/ModelBuilder depend on nightly build, in short, it's the project generated from CLI/ModelBuilder needs to depend on a released build of ML.Net

> All of Microsoft.ML.* should be self consistent with each other. No package pinning needed. Almost all changes to the AutoML code should have a mirroring change to the CodeGen code.
Hardly agree, there's no guarantee that every change in AutoML/ML.Net will be released, while projects generated by CodeGen should refer to a released version of ML.Net. If we mirroring every change in AutoML/ML.Net to CodeGen, that will make CodeGen in an unusable state between each release because ML.Net/AutoML CodeGen referred is different from what it's generated project referred.

> When I import Microsoft.ML.CodeGeneration@0.17.0 into my project, I'm expecting it to reference Microsoft.ML@1.5.0 & Microsoft.ML.AutoML@0.17.0. Lets say my project has Microsoft.ML@1.5.0 to do prediction, then I want to call CodeGen (yes, only an internal call currently). Then my project needs both Microsoft.ML@1.5.0 and Microsoft.ML@1.4.x, causing a version conflict. Consistent releases are better.

I don't see why it's related, if you want to call CodeGen, you need to call CodeGen that pin to `Microsoft.ML@1.5.0` instead of `Microsoft.ML@1.4.0`

In fact, if we can make sure before each release, CodeGen will be the final package to update, and during two releases, we don't make any changes to CodeGen, then consistent release works. But if during two releases there's any breaking change in API, then we must and mustn't update CodeGenerator because of the reason I mention above, either change or not change will make CodeGenerator in a breaking state.

File: test/Microsoft.ML.AutoML.Tests/ColumnInferenceTests.cs (hunk 5 above)
Comment (LittleLittleCloud) [reply test/Microsoft.ML.AutoML.Tests/ColumnInferenceTests.cs:215]: ### Resolved

File: src/Microsoft.ML.AutoML/ColumnInference/TextFileContents.cs (hunk 6 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/ColumnInference/TextFileContents.cs:28]: TextLoader Option uses ReadMultilines (So is allowQuote and allowSparse), and we'd better follow their naming style.

Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/ColumnInference/TextFileContents.cs:28]: ### Resolved
